include/zephyr/arch/arc/sys-io-common.h

```diff
@@ -74,6 +74,28 @@ static ALWAYS_INLINE void sys_write32(uint32_t data, mem_addr_t addr)
 	compiler_barrier();
 }
 
+static ALWAYS_INLINE uint64_t sys_read64(mem_addr_t addr)
```

Comment:
Not all ARC CPUs support atomic 64-bit load/store instructions, this would need some kconfigs and ifdef's (EM 32-bit only, some HS have 64-bit load/store, others don't depending on HW configuration).
Should a generic kconfig be introduced for this?
We already have CONFIG_64BIT to indicate 64-bit address space, but nothing for (atomic) 64bit load/stores.

Reply:
This is an interesting development.  This tends to indicate that the current state of sys_read64/sys_write64 is that some architectures *already* do not guarantee atomicity.  So the statements earlier by @andyross  about driver atomicity guarantees are not matched by the current implementation.

I'm fine with stating that the existing functions are not atomic and enabling more architectures.  Or if you feel strongly you can rename the existing accessor in ARC with a _nonatomic suffix.

@ruuddw  and @carlocaione can you make the call while @andyross is out?

Reply:
I think we should leave only ARM64 and X86 (? not sure about this) to have proper `sys_{read,write}64` functions and leave all the other platforms to use the `_nonatomic` suffix.

Reply:
My point was not that there are already non-atomic sys_read/write64 implementations in the tree, my point was specifically on the commit for ARC: it adds the (assumed to be atomic) functions to all ARC CPU family members, but that's not correct. Some support atomic access (as they are 64-bit architecture or have 64-bit load/store and bus interface), some don't.

My question/suggestion was to align on some kconfig option to indicate whether an CPU architecture supports atomic 64-bit load/stores or not, instead of having no or different approaches for that per architecture.

Reply:
With respect to 64 bit MMIO access on arm32 platforms, I will prefer sys_read64_nonatomic_.  as the atomicity is not guaranteed (https://lore.kernel.org/xen-devel/6D2FD203-04D9-4CDD-B57D-E90F5BC5CC3A@arm.com/T/#mcc2b226e506c669918eee0266067f6d746ccc744). This discussion came up as part of #51163

This is similar to what you wrote in https://github.com/zephyrproject-rtos/zephyr/pull/47171#issuecomment-1175254646.

Reply:
I believe adding _nonatomic is fine, and the atomic one only for platforms that support it. As that likely requires some #ifdef's in code, my suggestion was to introduce something for that in this PR, instead of leaving that to the individual architectures.